scp/scpInterpretationRequestInitiationAgent.cpp

```diff
@@ -0,0 +1,133 @@
+#include "scpInterpretationRequestInitiationAgent.hpp"
+
+#include "scpKeynodes.hpp"
+#include "scpUtils.hpp"
+
+#include <chrono>
+
+namespace scp
+{
+ScResult SCPInterpretationRequestInitiationAgent::DoProgram(ScElementaryEvent const & event, ScAction & action)
+{
+  auto const & startTime = std::chrono::high_resolution_clock::now();
+  sc_uint32 const maxCustomerWaitingTime = action.GetMaxCustomerWaitingTime();
+
+  ScAddr const & agentProgram = GetAgentProgram();
+  if (!m_context.IsElement(agentProgram))
+  {
+    SC_AGENT_LOG_ERROR("Agent program is not valid");
+    action.FinishUnsuccessfully();
+  }
+
+  ScAddr const & scpParams = m_context.GenerateNode(ScType::NodeConst);
+  ScAddr const & firstArgumentArc =
+      m_context.GenerateConnector(ScType::EdgeAccessConstPosPerm, scpParams, agentProgram);
+  m_context.GenerateConnector(ScType::EdgeAccessConstPosPerm, Keynodes::rrel_1, firstArgumentArc);
+
+  ScAddr const & secondArgumentArc =
+      m_context.GenerateConnector(ScType::EdgeAccessConstPosPerm, scpParams, event.GetConnector());
+  m_context.GenerateConnector(ScType::EdgeAccessConstPosPerm, Keynodes::rrel_2, secondArgumentArc);
+
+  ScAction scpAction = m_context.GenerateAction(Keynodes::action_scp_interpretation_request);
+  scpAction.SetArguments(agentProgram, scpParams);
+
+  ScAddr const & authorArc =
+      m_context.GenerateConnector(ScType::EdgeDCommonConst, scpAction, Keynodes::abstract_scp_machine);
+  m_context.GenerateConnector(ScType::EdgeAccessConstPosPerm, Keynodes::nrel_authors, authorArc);
+
+  if (maxCustomerWaitingTime == 0)
+    scpAction.InitiateAndWait();
+  else
+  {
+    auto const & timeFromStart = scp::Utils::GetTimeFromStart(startTime);
+    if (timeFromStart < maxCustomerWaitingTime)
+      scpAction.InitiateAndWait(maxCustomerWaitingTime - timeFromStart);
+    else
+    {
+      SC_AGENT_LOG_WARNING(
+          "Max customer waiting time"
+          << maxCustomerWaitingTime
+          << " has expired before action of class `action_scp_interpretation_request` was initiated because"
+          << timeFromStart << " ms have passed");
+      scpAction.InitiateAndWait();
+    }
+  }
+  auto const & resIt = m_context.CreateIterator5(
+      action, ScType::EdgeDCommonConst, ScType::NodeConst, ScType::EdgeAccessConstPosPerm, ScKeynodes::nrel_result);
+  if (resIt->Next())
```

Comment:
Should we log about that there are no result?

Reply:
I don't think that we should, it is not mandatory to have result